ibis/backends/bigquery/udf/__init__.py

```diff
@@ -313,5 +323,87 @@ def wrapped(*args, **kwargs):
         wrapped.js = js
         return wrapped
 
+    @staticmethod
+    def sql(
+        name: str,
+        params: Mapping[str, dt.DataType | Literal["ANY TYPE"]],
+        output_type: dt.DataType,
+        sql_expression: str,
+    ) -> Callable:
+        """Define a SQL UDF for BigQuery.
+
+        Parameters
+        ----------
+        name:
+            The name of the function.
+        params
+            Mapping of names and types of parameters
+        output_type
+            Return type of the UDF.
+        sql_expression
+            The SQL expression that defines the function.
+
+        Returns
+        -------
+        Callable
+            The wrapped user-defined function.
+
+        Examples
+        --------
+        >>> from ibis.backends.bigquery import udf
+        >>> import ibis.expr.datatypes as dt
+        >>> add_one = udf.sql(
+        ...    name="add_one",
+        ...    params={'x': dt.double},
+        ...    output_type=dt.double,
+        ...    sql_expression="x + 1"
+        ... )
+        >>> print(add_one.js)
+        CREATE TEMPORARY FUNCTION add_one_0(x FLOAT64)
+        RETURNS FLOAT64
+        AS (x + 1)
+        """
+        validate_output_type(output_type)
+        udf_node_fields = {
+            name: rlz.any if type_ == "ANY TYPE" else rlz.value(type_)
+            for name, type_ in params.items()
+        }
+
+        return_type = ibis_type_to_bigquery_type(dt.dtype(output_type))
+
+        udf_node_fields["output_dtype"] = output_type
+        udf_node_fields["output_shape"] = rlz.shape_like("args")
+        udf_node_fields["__slots__"] = ("js",)
+
+        udf_node = _create_udf_node(name, udf_node_fields)
+
+        from ibis.backends.bigquery.compiler import compiles
+
+        @compiles(udf_node)
+        def compiles_udf_node(t, op):
+            args_formatted = ", ".join(map(t.translate, op.args))
+            return f"{udf_node.__name__}({args_formatted})"
+
+        bigquery_signature = ", ".join(
+            "{name} {type}".format(
+                name=name,
+                type="ANY TYPE"
+                if type_ == "ANY TYPE"
+                else ibis_type_to_bigquery_type(dt.dtype(type_)),
+            )
+            for name, type_ in params.items()
+        )
+        js = f'''\
+CREATE TEMPORARY FUNCTION {udf_node.__name__}({bigquery_signature})
+RETURNS {return_type}
+AS ({sql_expression});'''
+
+        def wrapper(*args, **kwargs):
+            node = udf_node(*args, **kwargs)
+            object.__setattr__(node, "js", js)
```

Comment:
Should we call this something other than `js` now that we are generating both JavaScript and SQL? How about `code`?

Reply:
This attribute always contained SQL code, so it was always incorrectly named.

Reply:
Fixed.

Reply:
Ha! Fair point!